File: packages/java-sdk/src/test/java/com/turbodocx/ClientContextTest.java
package com.turbodocx;

import com.google.gson.JsonObject;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.*;

import java.io.IOException;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Client-context header tests (parity with JS tests/http-client-context.test.ts).
 *
 * The audit trail records device/location from request headers. The SDK must
 * send a descriptive User-Agent starting with "@turbodocx/sdk/", a timezone, a
 * language, an optional client IP (X-Forwarded-For -> geolocation), and a
 * device fingerprint. Uses OkHttp MockWebServer to capture the sent headers.
 */
class ClientContextTest {

    private MockWebServer server;

    @BeforeEach
    void setUp() throws IOException {
        server = new MockWebServer();
        server.start();
    }

    @AfterEach
    void tearDown() throws IOException {
        server.shutdown();
    }

    private HttpClient clientWith(ClientContext ctx) {
        return new HttpClient(server.url("/").toString(), "TDX-test-key", null, "test-org-id",
                "support@example.com", null, 60, 120, 60, ctx);
    }

    private RecordedRequest fireGet(ClientContext ctx) throws Exception {
        server.enqueue(new MockResponse().setHeader("Content-Type", "application/json").setBody("{\"data\":{\"ok\":true}}"));
        clientWith(ctx).get("/api/test", JsonObject.class);
        return server.takeRequest();
    }

    @Test
    void sendsDescriptiveTurboDocxSdkUserAgentByDefault() throws Exception {
        String ua = fireGet(ClientContext.autoDetect()).getHeader("User-Agent");
        assertTrue(ua.startsWith("@turbodocx/sdk/"), "got " + ua);
        assertFalse(ua.contains("okhttp"));
    }

    @Test
    void letsCallerOverrideUserAgent() throws Exception {
        RecordedRequest req = fireGet(ClientContext.builder().userAgent("my-app/9.9 (worker)").build());
        assertEquals("my-app/9.9 (worker)", req.getHeader("User-Agent"));
    }

    @Test
    void sendsAcceptLanguageFromHostLocaleByDefault() throws Exception {
        Locale original = Locale.getDefault();
        try {
            Locale.setDefault(Locale.forLanguageTag("en-US"));
            assertEquals("en-US", fireGet(ClientContext.autoDetect()).getHeader("Accept-Language"));
        } finally {
            Locale.setDefault(original);
        }
    }

    @Test
    void letsCallerOverrideLanguage() throws Exception {
        assertEquals("fr-FR", fireGet(ClientContext.builder().language("fr-FR").build()).getHeader("Accept-Language"));
    }

    @Test
    void letsCallerOverrideTimezone() throws Exception {
        assertEquals("America/New_York",
                fireGet(ClientContext.builder().timezone("America/New_York").build()).getHeader("X-Timezone"));
    }

    @Test
    void doesNotSendForwardedForByDefault() throws Exception {
        assertNull(fireGet(ClientContext.autoDetect()).getHeader("X-Forwarded-For"));
    }

    @Test
    void sendsForwardedForWhenCallerSuppliesIp() throws Exception {
        assertEquals("203.0.113.7",
                fireGet(ClientContext.builder().ipAddress("203.0.113.7").build()).getHeader("X-Forwarded-For"));
    }

    @Test
    void sendsDeviceFingerprintByDefaultAndHonorsOverride() throws Exception {
        assertNotNull(fireGet(ClientContext.autoDetect()).getHeader("X-Device-Fingerprint"));
        assertEquals("fp-abc",
                fireGet(ClientContext.builder().deviceFingerprint("fp-abc").build()).getHeader("X-Device-Fingerprint"));
    }

    @Test
    void preservesAuthAndOrgHeaders() throws Exception {
        RecordedRequest req = fireGet(ClientContext.autoDetect());
        assertEquals("Bearer TDX-test-key", req.getHeader("Authorization"));
        assertEquals("test-org-id", req.getHeader("x-rapiddocx-org-id"));
    }

    @Test
    void appliesContextHeadersOnMultipartUploads() throws Exception {
        server.enqueue(new MockResponse().setHeader("Content-Type", "application/json").setBody("{\"data\":{\"ok\":true}}"));
        Map<String, String> form = new HashMap<>();
        form.put("documentName", "x");
        clientWith(ClientContext.builder().ipAddress("203.0.113.7").build())
                .uploadFile("/turbosign/single/prepare-for-review", "%PDF-1.4 test".getBytes(), "doc.pdf", form, JsonObject.class);

        RecordedRequest req = server.takeRequest();
        assertTrue(req.getHeader("User-Agent").startsWith("@turbodocx/sdk/"));
        assertEquals("203.0.113.7", req.getHeader("X-Forwarded-For"));
        assertTrue(req.getHeader("Content-Type").contains("multipart/form-data"));
    }
}

File: packages/go-sdk/client_context_test.go
package turbodocx

import (
	"context"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/stretchr/testify/require"
)

// captureHeaders points an HTTPClient (built with the given client context) at a
// test server and returns the headers the server received for a GET request.
// This is the Go analog of the JS test inspecting the headers passed to fetch.
func captureHeaders(t *testing.T, ctx ClientContext) http.Header {
	t.Helper()
	var got http.Header
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got = r.Header.Clone()
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":{"ok":true}}`))
	}))
	defer server.Close()

	client := NewHTTPClient(ClientConfig{
		APIKey:        "TDX-test-key",
		OrgID:         "test-org-id",
		BaseURL:       server.URL,
		ClientContext: ctx,
	})
	var out map[string]interface{}
	require.NoError(t, client.Get(context.Background(), "/api/test", &out))
	return got
}

func TestSendsDescriptiveTurboDocxSDKUserAgentByDefault(t *testing.T) {
	ua := captureHeaders(t, ClientContext{}).Get("User-Agent")
	require.True(t, strings.HasPrefix(ua, "@turbodocx/sdk/"), "got %q", ua)
	require.NotContains(t, ua, "Go-http-client")
}

func TestLetsCallerOverrideUserAgent(t *testing.T) {
	ua := captureHeaders(t, ClientContext{UserAgent: "my-app/9.9 (worker)"}).Get("User-Agent")
	require.Equal(t, "my-app/9.9 (worker)", ua)
}

func TestSendsAcceptLanguageFromHostLocaleByDefault(t *testing.T) {
	t.Setenv("LC_ALL", "")
	t.Setenv("LC_MESSAGES", "")
	t.Setenv("LANG", "en_US.UTF-8")
	require.Equal(t, "en-US", resolveClientContextHeaders(ClientContext{})["Accept-Language"])
}

func TestLetsCallerOverrideLanguage(t *testing.T) {
	require.Equal(t, "fr-FR", captureHeaders(t, ClientContext{Language: "fr-FR"}).Get("Accept-Language"))
}

func TestLetsCallerOverrideTimezone(t *testing.T) {
	require.Equal(t, "America/New_York", captureHeaders(t, ClientContext{Timezone: "America/New_York"}).Get("X-Timezone"))
}

func TestDoesNotSendForwardedForByDefault(t *testing.T) {
	require.Empty(t, captureHeaders(t, ClientContext{}).Get("X-Forwarded-For"))
}

func TestSendsForwardedForWhenCallerSuppliesIP(t *testing.T) {
	require.Equal(t, "203.0.113.7", captureHeaders(t, ClientContext{IPAddress: "203.0.113.7"}).Get("X-Forwarded-For"))
}

func TestSendsDeviceFingerprintByDefaultAndHonorsOverride(t *testing.T) {
	require.NotEmpty(t, captureHeaders(t, ClientContext{}).Get("X-Device-Fingerprint"))
	require.Equal(t, "fp-abc", captureHeaders(t, ClientContext{DeviceFingerprint: "fp-abc"}).Get("X-Device-Fingerprint"))
}

func TestPreservesAuthOrgAndContentType(t *testing.T) {
	h := captureHeaders(t, ClientContext{})
	require.Equal(t, "Bearer TDX-test-key", h.Get("Authorization"))
	require.Equal(t, "test-org-id", h.Get("x-rapiddocx-org-id"))
	require.Equal(t, "application/json", h.Get("Content-Type"))
}

func TestAppliesContextHeadersOnMultipartUploads(t *testing.T) {
	var got http.Header
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got = r.Header.Clone()
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":{"ok":true}}`))
	}))
	defer server.Close()

	client := NewHTTPClient(ClientConfig{
		APIKey:        "TDX-test-key",
		OrgID:         "test-org-id",
		BaseURL:       server.URL,
		ClientContext: ClientContext{IPAddress: "203.0.113.7"},
	})
	var out map[string]interface{}
	require.NoError(t, client.UploadFile(context.Background(), "/turbosign/single/prepare-for-review",
		[]byte("%PDF-1.4 test"), "doc.pdf", map[string]string{"documentName": "x"}, &out))

	require.True(t, strings.HasPrefix(got.Get("User-Agent"), "@turbodocx/sdk/"))
	require.Equal(t, "203.0.113.7", got.Get("X-Forwarded-For"))
	// Multipart Content-Type must carry the multipart boundary, not application/json.
	require.Contains(t, got.Get("Content-Type"), "multipart/form-data")
}
